File: src/main/java/com/cae/mapped_exceptions/MappedException.java
package com.cae.mapped_exceptions;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class MappedException extends RuntimeException{

    protected final String briefPublicMessage;

    protected final String details;

    protected final Exception originalException;

    public MappedException(String briefPublicMessage, String details){
        super(briefPublicMessage + " | " + details);
        this.briefPublicMessage = briefPublicMessage;
        this.details = details;
        this.originalException = null;
    }

    public MappedException(String briefPublicMessage){
        super(briefPublicMessage);
        this.briefPublicMessage = briefPublicMessage;
        this.details = null;
        this.originalException = null;
    }

    public MappedException(String briefPublicMessage, String details, Exception originalException){
        super(briefPublicMessage + " | " + details + " | Original: " + originalException);
        this.briefPublicMessage = briefPublicMessage;
        this.details = details;
        this.originalException = originalException;
    }

    public MappedException(String briefPublicMessage, Exception originalException){
        super(briefPublicMessage + " | Original: " + originalException);
        this.briefPublicMessage = briefPublicMessage;
        this.details = null;
        this.originalException = originalException;
    }

    public String getBriefPublicMessage(){
        return this.briefPublicMessage;
    }

    /**
     * Getter method for the more detailed explanation of the exception.
     * @return Optional of the detailed message.
     */
    public Optional<String> getDetails() {
        return Optional.ofNullable(this.details);
    }

    public Optional<Exception> getOriginalException(){
        return Optional.ofNullable(this.originalException);
    }

    public List<String> getLinesFromStackTraceFromOriginalException(Integer numberOfLines){
        return Optional.ofNullable(this.originalException)
                .map(originalE -> this.getLinesFromStackTraceAsString(originalE, numberOfLines))
                .orElse(new ArrayList<>());
    }

    public List<String> getLinesFromStackTrace(Integer numberOfLines){
        return this.getLinesFromStackTraceAsString(this, numberOfLines);
    }

    public List<String> getLinesFromStackTraceAsString(Exception exception, Integer numberOfLines){
        var allLines = List.of(exception.getStackTrace());
        var linesToReturn = new ArrayList<String>();
        var counter = 0;
        var maxCount = allLines.size() < numberOfLines? allLines.size() : numberOfLines;
        while (counter < maxCount){
            var line = allLines.get(counter);
            linesToReturn.add(line.toString());
            counter ++;
        }
        if (allLines.size() > numberOfLines){
            var lastLine = allLines.size() - numberOfLines + " hidden line(s)";
            linesToReturn.add(lastLine);
        }
        return linesToReturn;
    }

    protected String getFullStackTraceAsString(){
        var stringWriter = new StringWriter();
        this.printStackTrace(new PrintWriter(stringWriter));
        return stringWriter.toString();
    }

}

File: implementations/python/src/cae_mapped_exceptions/mapped_exception.py
import traceback


class MappedException(Exception):

    def __init__(
            self,
            brief_public_message: str | None = None,
            details: str | None = None,
            original: Exception | None = None
    ):
        super().__init__(
            (f"{brief_public_message}") +
            (f" | Details: '{details}'" if details is not None else "") +
            (f" | Original: '{original}'" if original is not None else "")
        )
        self.brief_public_message = brief_public_message
        self.details = details
        self.original_exception = original

    @classmethod
    def with_brief_public_message(
            cls,
            brief_public_message: str,
            original_exception: Exception | None = None
    ):
        return cls(
            brief_public_message=brief_public_message,
            details=None,
            original=original_exception
        )

    @classmethod
    def with_full_details(
            cls,
            brief_public_message: str,
            details: str,
            original_exception: Exception | None = None
    ):
        return cls(
            brief_public_message=brief_public_message,
            details=details,
            original=original_exception
        )

    def get_original_exception_traceback_lines(
            self,
            limit: int) -> list[str]:
        if self.original_exception is not None:
            original_traceback = self.original_exception.__traceback__
            if original_traceback is None:
                return []
            lines = [line.rstrip("\n") for line in (
                traceback.format_tb(original_traceback)
            )]
            if len(lines) <= limit:
                return lines
            return lines[:limit] + [f"{len(lines) - limit} hidden line(s)"]
        return []
